Documentation/design-docs/struct-abi.md

```diff
@@ -0,0 +1,120 @@
+  requirements, which sets up the `fgArgInfo` and `argTable`, with the IR
+  transformations required to ensure that the arguments of the `GT_CALL` are
+  in the appropriate form.
+
+- When `fgCanFastTailCall()` is called, it doesn't yet have the `fgArgInfo`,
+  so it must duplicate some of the analysis that is done in `fgMorphArgs()`
+
+High-Level Proposed Design
+--------------------------
+This is a preliminary design, and is likely to change as the implementation proceeds:
+
+First, the `fgArgInfo` is extended to contain all the information needed to determine
+how an argument is passed. Ideally, most of the `#ifdef`s relating to ABI differences
+can be eliminated by querying the `fgArgInfo`. Most of the information will be queried
+via properties, such that when a target doesn't support a particular struct passing
+mechanism (e.g. passing structs by reference), the property will unconditionally return false, and the associated code paths will be eliminated.
+
+The initial determination of the number of arguments and how they
+are passed is extracted from `fgMorphArgs()` into a separate method: `gtInitArgInfo()`. It is idempotent - that is, it can be re-invoked and will simply return if it
+has already been called. It can be called by `fgCanFastTailCall()` so that it can query
+the `argTable` to get the information it requires.
+
+This method is responsible for the first part of what is currently `fgMorphArgs()`, plus setting up the `argTable`:
+- Count the number of args.
+  - Create any non-standard args that are needed, but don't yet create copies
+- Create the `argTable` for the given number of args
+- Initialize the `fgArgInfo` for each arg, with all the information about how
+  the arg is passed, and whether it requires a temp, but don't yet create any
+  temps.
+  - On x64/ux, this is the only method that should need to consult the struct
+    descriptor for outgoing arguments.
+  - The `isProcessed` flag remains false until `fgMorphArgs()` has handled
+    the arg.
+- Validate that each struct argument is either a `GT_LCL_VAR`, a `GT_OBJ`,
+  or a `GT_MKREFANY`.
+
+During the initial `fgMorph` phase, `fgMorphArgs()` does the following:
+
+- Calls `gtInitArgInfo()` to ensure that the `argTable` is set up properly.
+
+- Creates a copy of each argument as necessary.
+  - This should only be done if one or more of the following conditions hold:
+    - A copy is required to preserve possible ordering dependencies, in which
+      case the `needsTmp` field of the `fgArgInfo` was set to true by
+      `fgInitArgInfo()`.
+    - A struct arg has been promoted, it is passed in register(s) (or split),
+      and has not yet been marked `lvDoNotEnregister`.
+
+- Sets up the actual argument for any non-standard args.
+
+- Transform struct arg nodes from `GT_LCL_VAR`, `GT_OBJ` or `GT_MKREFANY` into:
+  - `GT_FIELD_LIST` (i.e. a list of fields) if the lclVar is promoted and
```

Comment:
Wouldn't it be possible to avoid the need for this field list concept by "decomposing" the call argument into multiple arguments of primitive types? That is, if you have a struct containing 3 `int` fields and the calling convention requires it to be passed on the stack can't that single struct argument be split into 3 separate `int` arguments (each initialized from the 3 lclvars that were supposed to be in the field list) ?

Reply:
I *think* that would add more complexity than simplification. I don't think we'd want to do it up-front (i.e. at the time of the proposed `fgInitArgInfo()`, as it would unnecessarily break up the codegen for the case where it's loaded from the stack. Decomposing later would mean that we'd either have to change the `fgArgInfo` (making it not match the actual signature), or have some other mechanism to map back from these "partial" arguments to their associated `fgArgInfo`. There may be a simpler approach, but it's not evident to me.
That said, I suspect (as I mention in the document) that the design will evolve over time as we gain a better understanding of how well it plays with the rest of the JIT.

Reply:
>  as it would unnecessarily break up the codegen for the case where it's loaded from the stack. 

I'm not sure I understand this part. In any case, if it cannot be done up-front then it's probably not worth doing. The reason I think such decomposition might be useful is that it may allow more node ordering flexibility, I've seen this happening with long arguments in my long helper call experiments. And of course, it would avoid this concept of "field list" but then I'm not sure it can avoid it completely for this advantage to be relevant.

> Decomposing later would mean that we'd either have to change the fgArgInfo (making it not match the actual signature), or have some other mechanism to map back from these "partial" arguments to their associated fgArgInfo.

Yes, if done I'd think it should be reflected in arg info. The call node should look exactly as if it was a call with 3 `int` arguments, except differences required by the ABI (e.g. passing those args on stack instead of regs). And yes, it would be different from the actual signature, does that matter? I kind of expect it not to be needed after import/morph.

Reply:
Don't forget cases where multiple struct fields are passed in a single register and need to be packed/unpacked to/from multiple promoted registers.

Reply:
> Don't forget cases where multiple struct fields are passed in a single register and need to be packed/unpacked to/from multiple promoted registers.

Right, and it's not only about fields passed in registers, it also affects structs passed on the stack if multiple fields end up in the same stack slot. So it's not really about "decomposing" arguments into primitive type fields/args but rather decomposing into registers and stack slots. Unfortunately that would likely be way too complicated to do as it also requires generating IR to pack such fields.